File: test/language/optional_members/java/optional_members/OptionalArrayRecursionTest.java
package optional_members;

import static org.junit.Assert.*;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.junit.Test;

import optional_members.optional_array_recursion.Employee;
import optional_members.optional_array_recursion.Title;

import zserio.runtime.array.ObjectArray;
import zserio.runtime.io.FileBitStreamReader;

public class OptionalArrayRecursionTest
{
    @Test
    public void bitSizeOf()
    {
        final Employee employee = createEmployee(EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title.DEVELOPER);
        assertEquals(DEVELOPER1_BIT_SIZE, employee.bitSizeOf());

        final Employee teamLead = createTeamLead();
        assertEquals(TEAM_LEAD_BIT_SIZE, teamLead.bitSizeOf());
    }

    @Test
    public void hasTeamMembers()
    {
        final Employee employee = createEmployee(EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title.DEVELOPER);
        assertFalse(employee.hasTeamMembers());

        final Employee teamLead = createTeamLead();
        assertTrue(teamLead.hasTeamMembers());
    }

    @Test
    public void equals()
    {
        final Employee emptyEmployee1 = new Employee();
        final Employee emptyEmployee2 = new Employee();
        assertTrue(emptyEmployee1.equals(emptyEmployee2));

        final Employee teamLead1 = createTeamLead();
        assertFalse(teamLead1.equals(emptyEmployee1));

        final Employee teamLead2 = createTeamLead();
        assertTrue(teamLead1.equals(teamLead2));
    }

    @Test
    public void hashCodeMethod()
    {
        final Employee emptyEmployee1 = new Employee();
        final Employee emptyEmployee2 = new Employee();
        assertEquals(emptyEmployee1.hashCode(), emptyEmployee2.hashCode());

        final Employee teamLead1 = createTeamLead();
        assertTrue(teamLead1.hashCode() != emptyEmployee1.hashCode());

        final Employee teamLead2 = createTeamLead();
        assertEquals(teamLead1.hashCode(), teamLead2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        final Employee employee = createEmployee(EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title.DEVELOPER);
        final int bitPosition = 1;
        assertEquals(bitPosition + DEVELOPER1_BIT_SIZE, employee.initializeOffsets(bitPosition));

        final Employee teamLead = createTeamLead();
        assertEquals(bitPosition + TEAM_LEAD_BIT_SIZE, teamLead.initializeOffsets(bitPosition));
    }

    @Test
    public void fileWriteEmployee() throws IOException
    {
        final Employee employee = createEmployee(EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title.DEVELOPER);

        final File employeeFile = new File("employee.bin");
        employee.write(employeeFile);
        final FileBitStreamReader reader = new FileBitStreamReader(employeeFile);
        checkEmployeeInStream(reader, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY, Title.DEVELOPER);
        reader.close();

        Employee readEmployee = new Employee(employeeFile);
        assertEquals(EMPLOYEE_DEVELOPER1_NAME, readEmployee.getName());
        assertEquals(EMPLOYEE_DEVELOPER1_SALARY, readEmployee.getSalary());
        assertEquals(Title.DEVELOPER, readEmployee.getTitle());
    }

    @Test
    public void fileWriteTeamLead() throws IOException
    {
        final Employee teamLead = createTeamLead();

        final File teamLeadFile = new File("team_lead.bin");
        teamLead.write(teamLeadFile);
        final FileBitStreamReader reader = new FileBitStreamReader(teamLeadFile);
        checkTeamLeadInStream(reader);
        reader.close();

        Employee readTeamLead = new Employee(teamLeadFile);
        assertEquals(EMPLOYEE_TEAM_LEAD_NAME, readTeamLead.getName());
        assertEquals(EMPLOYEE_TEAM_LEAD_SALARY, readTeamLead.getSalary());
        assertEquals(Title.TEAM_LEAD, readTeamLead.getTitle());
        assertEquals(NUM_DEVELOPERS, readTeamLead.getTeamMembers().length());
    }

    private static Employee createEmployee(String name, int salary, Title title)
    {
        final Employee employee = new Employee();
        employee.setName(name);
        employee.setSalary(salary);
        employee.setTitle(title);

        return employee;
    }

    private static Employee createTeamLead()
    {
        final Employee teamLead = createEmployee(EMPLOYEE_TEAM_LEAD_NAME, EMPLOYEE_TEAM_LEAD_SALARY,
                Title.TEAM_LEAD);

        final List<Employee> teamMembers = new ArrayList<Employee>();
        final Employee teamMember1 = createEmployee(EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title.DEVELOPER);
        teamMembers.add(teamMember1);
        final Employee teamMember2 = createEmployee(EMPLOYEE_DEVELOPER2_NAME, EMPLOYEE_DEVELOPER2_SALARY,
                Title.DEVELOPER);
        teamMembers.add(teamMember2);
        teamLead.setTeamMembers(new ObjectArray<Employee>(teamMembers));

        return teamLead;
    }

    private static void checkEmployeeInStream(FileBitStreamReader reader, String name, int salary,
            Title title) throws IOException
    {
        assertEquals(name, reader.readString());
        assertEquals(salary, reader.readBits(16));
        assertEquals(title.getValue(), reader.readBits(8));
    }

    private static void checkTeamLeadInStream(FileBitStreamReader reader) throws IOException
    {
        checkEmployeeInStream(reader, EMPLOYEE_TEAM_LEAD_NAME, EMPLOYEE_TEAM_LEAD_SALARY, Title.TEAM_LEAD);
        assertEquals(NUM_DEVELOPERS, reader.readVarUInt64());
        checkEmployeeInStream(reader, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY, Title.DEVELOPER);
        checkEmployeeInStream(reader, EMPLOYEE_DEVELOPER2_NAME, EMPLOYEE_DEVELOPER2_SALARY, Title.DEVELOPER);
    }

    private static String   EMPLOYEE_TEAM_LEAD_NAME = "Nico";
    private static int      EMPLOYEE_TEAM_LEAD_SALARY = 2000;

    private static String   EMPLOYEE_DEVELOPER1_NAME = "Mike";
    private static int      EMPLOYEE_DEVELOPER1_SALARY = 1000;

    private static String   EMPLOYEE_DEVELOPER2_NAME = "Luke";
    private static int      EMPLOYEE_DEVELOPER2_SALARY = 1800;

    private static int      NUM_DEVELOPERS = 2;
    private static int      DEVELOPER1_BIT_SIZE = EMPLOYEE_DEVELOPER1_NAME.length() * 8 + 32;
    private static int      TEAM_LEAD_BIT_SIZE = EMPLOYEE_TEAM_LEAD_NAME.length() * 8 + 32 + 8 +
            DEVELOPER1_BIT_SIZE + EMPLOYEE_DEVELOPER2_NAME.length() * 8 + 32;
}

File: test/language/optional_members/cpp98/OptionalArrayRecursionTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#
#include "optional_members/optional_array_recursion/Employee.h"
#include "optional_members/optional_array_recursion/Title.h"

namespace optional_members
{
namespace optional_array_recursion
{

class OptionalArrayRecursionTest : public ::testing::Test
{
protected:
    void fillEmployee(Employee& employee, const char name[], uint16_t salary, Title title)
    {
        employee.setName(name);
        employee.setSalary(salary);
        employee.setTitle(title);
    }

    void fillTeamLead(Employee& teamLead)
    {
        fillEmployee(teamLead, EMPLOYEE_TEAM_LEAD_NAME, EMPLOYEE_TEAM_LEAD_SALARY, Title::TEAM_LEAD);

        zserio::ObjectArray<Employee> teamMembers;
        Employee teamMember1;
        fillEmployee(teamMember1, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY, Title::DEVELOPER);
        teamMembers.push_back(teamMember1);

        Employee teamMember2;
        fillEmployee(teamMember2, EMPLOYEE_DEVELOPER2_NAME, EMPLOYEE_DEVELOPER2_SALARY, Title::DEVELOPER);
        teamMembers.push_back(teamMember2);

        teamLead.setTeamMembers(teamMembers);
    }

    void checkEmployeeInBitStream(zserio::BitStreamReader& reader, const char name[], uint16_t salary,
            Title title)
    {
        ASSERT_EQ(name, reader.readString());
        ASSERT_EQ(salary, reader.readBits(16));
        ASSERT_EQ(title, reader.readBits(8));
    }

    void checkTeamLeadInBitStream(zserio::BitStreamReader& reader)
    {
        checkEmployeeInBitStream(reader, EMPLOYEE_TEAM_LEAD_NAME, EMPLOYEE_TEAM_LEAD_SALARY, Title::TEAM_LEAD);
        ASSERT_EQ(NUM_DEVELOPERS, reader.readVarUInt64());
        checkEmployeeInBitStream(reader, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY,
                Title::DEVELOPER);
        checkEmployeeInBitStream(reader, EMPLOYEE_DEVELOPER2_NAME, EMPLOYEE_DEVELOPER2_SALARY,
                Title::DEVELOPER);
    }

    static const char   EMPLOYEE_TEAM_LEAD_NAME[];
    static uint16_t     EMPLOYEE_TEAM_LEAD_SALARY;

    static const char   EMPLOYEE_DEVELOPER1_NAME[];
    static uint16_t     EMPLOYEE_DEVELOPER1_SALARY;

    static const char   EMPLOYEE_DEVELOPER2_NAME[];
    static uint16_t     EMPLOYEE_DEVELOPER2_SALARY;

    static const size_t NUM_DEVELOPERS;
    static const size_t EMPTY_EMPLOYEE_BIT_SIZE;
    static const size_t TEAM_LEAD_BIT_SIZE;
};

const char  OptionalArrayRecursionTest::EMPLOYEE_TEAM_LEAD_NAME[] = "Nico";
uint16_t    OptionalArrayRecursionTest::EMPLOYEE_TEAM_LEAD_SALARY = 2000;

const char  OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER1_NAME[] = "Mike";
uint16_t    OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER1_SALARY = 1000;

const char  OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER2_NAME[] = "Luke";
uint16_t    OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER2_SALARY = 1800;

const size_t OptionalArrayRecursionTest::NUM_DEVELOPERS = 2;
const size_t OptionalArrayRecursionTest::EMPTY_EMPLOYEE_BIT_SIZE = 32;
const size_t OptionalArrayRecursionTest::TEAM_LEAD_BIT_SIZE = EMPTY_EMPLOYEE_BIT_SIZE * 3 + 8 +
        (sizeof(OptionalArrayRecursionTest::EMPLOYEE_TEAM_LEAD_NAME) - 1) * 8 +
        (sizeof(OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER1_NAME) - 1) * 8 +
        (sizeof(OptionalArrayRecursionTest::EMPLOYEE_DEVELOPER2_NAME) - 1) * 8;

TEST_F(OptionalArrayRecursionTest, hasTeamMembers)
{
    Employee teamLead;
    ASSERT_FALSE(teamLead.hasTeamMembers());

    fillTeamLead(teamLead);
    ASSERT_TRUE(teamLead.hasTeamMembers());
}

TEST_F(OptionalArrayRecursionTest, bitSizeOf)
{
    Employee teamLead;
    ASSERT_EQ(EMPTY_EMPLOYEE_BIT_SIZE, teamLead.bitSizeOf());

    fillTeamLead(teamLead);
    ASSERT_EQ(TEAM_LEAD_BIT_SIZE, teamLead.bitSizeOf());
}

TEST_F(OptionalArrayRecursionTest, initializeOffsets)
{
    Employee teamLead;
    const size_t bitPosition = 1;
    ASSERT_EQ(bitPosition + EMPTY_EMPLOYEE_BIT_SIZE, teamLead.initializeOffsets(bitPosition));

    fillTeamLead(teamLead);
    ASSERT_EQ(bitPosition + TEAM_LEAD_BIT_SIZE, teamLead.initializeOffsets(bitPosition));
}

TEST_F(OptionalArrayRecursionTest, operatorEquality)
{
    Employee teamLead1;
    Employee teamLead2;
    ASSERT_TRUE(teamLead1 == teamLead2);

    fillTeamLead(teamLead1);
    ASSERT_FALSE(teamLead1 == teamLead2);

    fillTeamLead(teamLead2);
    ASSERT_TRUE(teamLead1 == teamLead2);
}

TEST_F(OptionalArrayRecursionTest, hashCode)
{
    Employee teamLead1;
    Employee teamLead2;
    ASSERT_EQ(teamLead1.hashCode(), teamLead2.hashCode());

    fillTeamLead(teamLead1);
    ASSERT_NE(teamLead1.hashCode(), teamLead2.hashCode());

    fillTeamLead(teamLead2);
    ASSERT_EQ(teamLead1.hashCode(), teamLead2.hashCode());
}

TEST_F(OptionalArrayRecursionTest, writeEmployee)
{
    Employee employee;
    fillEmployee(employee, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY, Title::DEVELOPER);

    zserio::BitStreamWriter writer;
    employee.write(writer);
    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    checkEmployeeInBitStream(reader, EMPLOYEE_DEVELOPER1_NAME, EMPLOYEE_DEVELOPER1_SALARY, Title::DEVELOPER);
    reader.setBitPosition(0);

    Employee readTeamLead(reader);
    ASSERT_EQ(EMPLOYEE_DEVELOPER1_NAME, readTeamLead.getName());
    ASSERT_EQ(EMPLOYEE_DEVELOPER1_SALARY, readTeamLead.getSalary());
    ASSERT_EQ(Title::DEVELOPER, readTeamLead.getTitle());
}

TEST_F(OptionalArrayRecursionTest, writeTeamLead)
{
    Employee teamLead;
    fillTeamLead(teamLead);

    zserio::BitStreamWriter writer;
    teamLead.write(writer);
    size_t writerBufferByteSize;
    const uint8_t* writerBuffer = writer.getWriteBuffer(writerBufferByteSize);
    zserio::BitStreamReader reader(writerBuffer, writerBufferByteSize);
    checkTeamLeadInBitStream(reader);
    reader.setBitPosition(0);

    Employee readTeamLead(reader);
    ASSERT_EQ(EMPLOYEE_TEAM_LEAD_NAME, readTeamLead.getName());
    ASSERT_EQ(EMPLOYEE_TEAM_LEAD_SALARY, readTeamLead.getSalary());
    ASSERT_EQ(Title::TEAM_LEAD, readTeamLead.getTitle());
    ASSERT_TRUE(readTeamLead.hasTeamMembers());
    zserio::ObjectArray<Employee> teamMembers = readTeamLead.getTeamMembers();
    ASSERT_EQ(NUM_DEVELOPERS, teamMembers.size());
}

} // namespace optional_array_recursion
} // namespace optional_members
